Writing the first version of the spec

sharnoff opened the pull request:
This pull request should remain open until the specification is deemed satisfactory for an initial implementation.

sharnoff committed:
initial progress on spec

sharnoff committed:
partial work on string literals

sharnoff committed:
(mostly) finished literals section

JOT85 committed:
Added concurrency document with a few ideas in

JOT85 committed:
Made some changes to spec: added basic tuple and block syntax, fn
declarations, and made the comment spec less ambiguous - I personally
think that the first `*/` should close the comment... But if you're
trying to mimic rust, we need to have a conversation, because rust has
some funky rules!!!!!!

JOT85 committed:
Added aims document, feel free to add anything else you think of! Or
refactor.

JOT85 committed:
Add a newline between the last 2 lines...

JOT85 committed:
That looked eww...

JOT85 commented:
Are we adding anything about formal proofs to the aims?

JOT85 committed:
I almost forgot it was that trivial!

JOT85 committed:
Updated README

sharnoff commented on spec.md:
````diff
@@ -0,0 +1,246 @@
+### Fn Declaration
+A fn declaration is used to define a function, for example:
+```
+  fn f(x) => ...  (where ... is an `expression`)
````
I think there are very few low-level languages that allow function definitions to infer their types. Is there any benefit from doing this? It seems like a no-go for public functions, is there a use case for private ones?

JOT85 replied:
I think that makes sense for public functions, since their interface matters so making a change to it should be explicit. Plus, the inferred types might be looser then intended, in a way in which the code will works, but the type doesn't tell you about the intention of the function.
For private functions, I think the same logic applies, however I think it's neater if we allow a lambda functions type to be inferred?
Public variables should be explicitly typed in the low level too, which would mean that a public lambda function would need to have its types specified.

sharnoff committed:
Added to fn declarations

sharnoff commented on spec.md:
```diff
@@ -87,6 +87,7 @@ Here is a table of keywords and their uses:
 | Keyword   | Type         | Uses                             |
 |-----------|--------------|----------------------------------|
 | **fn**    | Declaration  | [`FnDecl`](#fn-declaration)      |
+| **let**   | Declaration  | [`LetDecl`](#let-declaration)    |
```
I believe `let` is an expression, no? It returns a boolean

sharnoff commented on spec.md:
```diff
@@ -73,8 +73,8 @@ octal_digit = "0" … "7" .
-  1. *Line comments* (`LineComment`s) start with the sequence `//` and end at the end of the line.
-  2. *Block comments* (`BlockComment`s) start with the sequence `/*` and end with a matching `*/`.
+  1. *Line comments* (`LineComment`s) start with the sequence `//` and end at the next `newline`.
+  2. *Block comments* (`BlockComment`s) start with the sequence `/*` and end with next occurrence of `*/`.
```
Block comments should end on a *matching* `*/` because it's something that's very common in many "modern" languages - it's not just Rust. Scala, Swift, and Dart all do, as well. It seems to me that Go's the exception among recent languages in the way that it doesn't allow this.

JOT85 replied:
I made a comment in my commit message about this. I know it's a feature in a few languages, but I'm not sure what I think of it.
If we do specify it to be `matching`, we need to be careful about how we do it. Rusts behaviour isn't as you'd expect.
Perhaps we allow comments to be nested? So // hello */ wouldn't close it, since the */ is in another comment.
Also, out there suggestion, why don't we require the opening and closing things to have a matching number of *s. Most people wouldn't know this at first, potentially making it confusing, but I think it's a nice feature to have, since sometimes, it is convenient to not have nested comments.
I'm not sure what I think to that idea because it seems like a deviation without much benefit, but just to put it out there.

JOT85 committed:
Added more detail to aims.md

JOT85 committed:
Changed let from `declaration` to `expression`

JOT85 committed:
Added a brief list of things that the standard library should consist
of, and a possible structure.
Note that we'll probably want to prefix all of the non-core modules with
some trenchcoat organization prefix once we've decided upon how we're
doing that.
